Complete hidden commands

rcorre opened the pull request:
Note that with the refactoring in 52d0735, bindings are now included in the help completion model but not shown as it allocates 0 width for the `misc` column. Doing so can clip some setting descriptions so I wasn't sure if it was worth it.

Resolves #1707

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -257,3 +239,47 @@ def delete_cur_item(self, completion):
+    for obj in set(cmdutils.cmd_dict.values()):
+        if ((not obj.debug or objreg.get('args').debug) and
+             not obj.deprecated and
+             (include_hidden or not obj.hide)):
```
This is getting too long :wink:

What about:

```python
        hide_debug = obj.debug and not objreg.get('args').debug
        hide_hidden = obj.hide and not include_hidden
        if not (hide_debug or hide_hidden or obj.deprecated):
```

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -39,23 +39,11 @@ class CommandCompletionModel(base.BaseCompletionModel):
-        for (name, desc) in sorted(cmdlist):
-            self.new_item(cat, name, desc, ', '.join(cmd_to_keys[name]))
+        for (name, desc, misc) in cmdlist:
+            self.new_item(cat, name, desc, misc, )
```
I guess that trailing comma/space is unintended? :wink:

rcorre replied:
Didn't even know that was legal :O

The-Compiler replied:
Neither did I!

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -257,3 +239,47 @@ def delete_cur_item(self, completion):
+        if ((not obj.debug or objreg.get('args').debug) and
+             not obj.deprecated and
+             (include_hidden or not obj.hide)):
+            bindings = ', '.join(cmd_to_keys[obj.name])
```
I was really confused why this worked and didn't raise `KeyError` for unbound commands - I think it does before `get_reverse_bindings_for` uses a `collections.defaultdict(list)` to build up things and then returns that.

I think it would be cleaner for `get_reverse_bindings_for` to return a normal `dict`, and then we'd have explicitly do `cmd_to_keys.get(obj.name, [])` here. What do you think?

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -257,3 +239,47 @@ def delete_cur_item(self, completion):
+
+    if include_aliases:
+        for name, cmd in config.section('aliases').items():
+            cmdlist.append((name, "Alias for '{}'".format(cmd), ''))
```
I guess you could include keybindings to an alias here too?

The-Compiler commented on tests/unit/completion/test_models.py:
```diff
@@ -447,3 +449,34 @@ def test_setting_value_completion(qtmodeltester, monkeypatch, stubs,
+
+def test_bind_completion(qtmodeltester, monkeypatch, stubs, config_stub,
+                         key_config_stub):
+    """Test the results of command completion.
```
command -> bind?

The-Compiler commented on qutebrowser/config/parsers/keyconf.py:
```diff
@@ -422,7 +422,7 @@ def get_bindings_for(self, section):
     def get_reverse_bindings_for(self, section):
         """Get a dict of commands to a list of bindings for the section."""
-        cmd_to_keys = collections.defaultdict(list)
+        cmd_to_keys = {}
```
This probably won't work due to the `cmd_to_keys[cmd].append(key)` below, no? I was thinking about still using a `collections.defaultdict`, but then doing `return dict(cmd_to_keys)` to convert it to a normal dict.

The-Compiler commented:
What's the state of this, @rcorre? :wink:

rcorre commented:
Whoops, forgot I had this open. I'll see what's possible with showing bindings only if they are bound to a particular alias (and not the aliased command) and try to figure out why I broke the download tests.

rcorre commented:
Ok, so I threw in one more commit in this area. I didn't see any good reason that the completion models should sort the command list as the `sortfilter` will sort it anyways, so I removed it in b6abd14. I also refactored the tests to ignore the order of completion items.

The-Compiler commented on qutebrowser/config/parsers/keyconf.py:
```diff
@@ -422,8 +422,9 @@ def get_bindings_for(self, section):
-        cmd_to_keys = collections.defaultdict(list)
+        cmd_to_keys = {}
         for key, cmd in self.get_bindings_for(section).items():
+            cmd_to_keys.setdefault(cmd, [])
```
Nice, I didn't know about `dict.setdefault`!

The-Compiler commented on tests/unit/completion/test_models.py:
```diff
@@ -31,17 +31,18 @@
+    """Check that a model contains the expected items in any order.
+
+    Args:
+        expected: A list of form
```
That sure is a weird list :wink: I can change it to "dict" while merging though if nothing else comes up.

The-Compiler commented on tests/unit/completion/test_models.py:
```diff
@@ -50,8 +51,12 @@ def _get_completions(model):
+        assert cat_name in actual
+        actual_items = actual[cat_name]
+        for expected_item in expected_entries:
+            assert any(item for item in actual_items if item == expected_item)
```
I *think* this does the right thing because `any([])` is `False`, and `item` always would be truthy (as it's a non-empty list).

But wouldn't `assert expected_item in actual_items` do the same?

rcorre replied:
uh, yeah. Don't know how I came up with something so complicated :confounded:

The-Compiler commented:
Thanks!
